master/internal/api_master.go

```diff
@@ -25,3 +26,41 @@ func (a *apiServer) GetMaster(
 		Config:    configStruct,
 	}, err
 }
+
+func (a *apiServer) MasterLogs(
+	req *apiv1.MasterLogsRequest, resp apiv1.Determined_MasterLogsServer) error {
+	if err := grpc.ValidateRequest(
+		func() (bool, string) { return req.Limit >= 0, "Limit must be >= 0" },
+	); err != nil {
+		return err
+	}
+	total := a.m.logs.Len()
+	offset := int(req.Offset)
+	if req.Offset < 0 {
+		offset = total + offset
+		if offset < 0 {
+			offset = 0
+		}
+	}
+
+	limit := -1
+	if req.Limit != 0 {
+		limit = int(req.Limit)
+		if limit > total-offset {
+			limit = total - offset
+		}
+	}
```

Comment:
non-blocking: can we pull both of these blocks into helper functions that help explain what's going on (by having nice function names)?